### PR TITLE
Kurtwheeler sentry

### PR DESCRIPTION
**Pull request by kurtwheeler**
## Issue Number

#317 

## Purpose/Implementation Notes

We want to use sentry to record and track errors and bugs! This PR adds an integration to do so!

 This PR adds three different mechanisms which send errors to https://sentry.io/greenelab/refinebio/:
  * Any log message with level WARNING or higher will also be sent to sentry via an additional log handler configured in the `common` project.
  * Any uncaught exception will be sent to sentry: https://raven.readthedocs.io/en/stable/integrations/django.html#setup
  * Any request our API responds with a status code 400 or higher will cause a message to be sent to sentry 5% of the time so we can watch trends without having a deluge of issues come from the API.

## Types of changes

- New feature (non-breaking change which adds functionality)

## Functional tests

I have triggered errors in the workers and foreman projects by calling `raise ValueError()` within a `try` block so it would be logged and that log message would be sent to sentry. **WORKS!**

I put `1 / 0` at the top of a management command to make sure that uncaught exceptions get sent to sentry. **WORKS!**

I modified the ListSamples endpoint to raise a 403 error to make sure it made it into Sentry. **WORKS!**

I have triggered errors that should be sent to sentry but without the `RAVEN_DSN` env var being set and made sure that Raven didn't complain in any way about not being configured. **WORKS**

## Checklist

- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

## Screenshots

sentry_issues_list

sentry_403

### CODE REVIEW

**Comment by Miserlou** [comment]
Is there any way to test this? If this introduces new middleware that all of our API responses will be going through, I'd really like to make sure that there is a well-covered test. Looks good otherwise.

File: api/data_refinery_api/middleware.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+        from raven.contrib.django.models import client
+
+        if not client.is_enabled():
+            return response
+
+        data = client.get_data_from_request(request)
+        data.update({
+            "level": logging.WARN,
+            "logger": "BadRequestCatcher?",
```
Comment (Miserlou) [inline api/data_refinery_api/middleware.py]: What does the `?` in `BadRequestCatcher?` mean?

File: api/data_refinery_api/settings.py [hunk 2]
```diff
@@ -159,6 +161,25 @@
...
+# /usr/local/lib/python3.6/site-packages/raven/conf/remote.py:91:
+# UserWarning: Transport selection via DSN is deprecated. You should
+# explicitly pass the transport class to Client() instead.
+raven_dsn = get_env_variable('RAVEN_DSN', "not set")
+if raven_dsn != "not set":
+    RAVEN_CONFIG = {
+        'dsn': raven_dsn,
+        # Only send 5% of errors for the API, since we aren't going to
+        # be interested in any single one.
+        'sampleRate': 0.05
```
Comment (Miserlou) [inline api/data_refinery_api/settings.py]: Let's crank this to 15/25 for the first release then lower it, I don't want to miss anything while we're still low traffic.

File: common/data_refinery_common/logging.py [hunk 3]
```diff
@@ -41,4 +41,15 @@ def get_and_configure_logger(name: str) -> logging.Logger:
         fmt=FORMAT_STRING, keywords=[]))
     logger.logger.addHandler(handler)
 
+    # This is the Sentry handler
+    raven_dsn = get_env_variable("RAVEN_DSN", "not set")
```
Comment (Miserlou) [inline common/data_refinery_common/logging.py]: String matching here is a bit weird - we can't just use False?

File: foreman/data_refinery_foreman/settings.py [hunk 4]
```diff
@@ -140,3 +141,20 @@
...
 
 STATIC_URL = '/static/'
+
+
+# Setting the RAVEN_CONFIG when RAVEN_DSN isn't set will cause the
+# following warning:
+# /usr/local/lib/python3.6/site-packages/raven/conf/remote.py:91:
+# UserWarning: Transport selection via DSN is deprecated. You should
+# explicitly pass the transport class to Client() instead.
+raven_dsn = get_env_variable('RAVEN_DSN', "not set")
```
Comment (Miserlou) [inline foreman/data_refinery_foreman/settings.py]: Same comment as above

File: workers/data_refinery_workers/settings.py [hunk 5]
```diff
@@ -145,3 +139,20 @@ def get_env_variable(var_name):
...
 
 STATIC_URL = '/static/'
+
+
+# Setting the RAVEN_CONFIG when RAVEN_DSN isn't set will cause the
+# following warning:
+# /usr/local/lib/python3.6/site-packages/raven/conf/remote.py:91:
+# UserWarning: Transport selection via DSN is deprecated. You should
+# explicitly pass the transport class to Client() instead.
+raven_dsn = get_env_variable('RAVEN_DSN', "not set")
```
Comment (Miserlou) [inline workers/data_refinery_workers/settings.py]: And again

**Comment by Miserlou** [comment]
ACTUALLY now that I'm using Sentry, is there anyway that we can use different settings for API and Crunch endpoints? They're all mixed together.

**Comment by Miserlou** [comment]
It looks like the API responses aren't getting the correct remote IP address, it returns `"172.17.0.1"` which is the docker bridge network.

**Comment by kurtwheeler** [comment]
You're definitely correct about that IP. I've now spent quite a bit of time looking into this. It seems like this actually isn't the most simple issue to resolve. There's an issue that's almost three years old about it: https://github.com/moby/moby/issues/15086 and another that's close to two years old: https://github.com/docker/for-mac/issues/180 (it's not actually specific to mac)

I was able to get the actual IP by using the `--net=host \` option that was suggested in that issue. However this conflicts with the `--link drdb:postgres` flag and perhaps has some other implications. I'm not exactly sure what this would do to the prod configuration, if anything. Therefore I think I'll wait till we can discuss this in person on Monday to make sure that there's nothing else that would need to be done along with that change.

In the meantime I'm going to merge this PR so that I can kick off some crunch with Sentry enabled. This remaining issue is only going to affect the API, and we don't have too many users there anyway. I'm opening an issue to keep track of this IP issue.

**Comment by kurtwheeler** [comment]
I've made a new issue to track the IP issue here: https://github.com/AlexsLemonade/refinebio/issues/368

**Pull request merged**